logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/MessageSender.kt

```diff
@@ -243,31 +297,89 @@ internal class MessageSenderImpl internal constructor(
             .fold({
-                when (it) {
-
-                    is ProteusSendMessageFailure -> {
-                        logger.w("Proteus Send Failure: { \"message\" : \"${message.toLogString()}\", \"errorInfo\" : \"${it}\" }")
-                        messageSendFailureHandler
-                            .handleClientsHaveChangedFailure(it)
-                            .flatMap {
-                                logger.w("Retrying After Proteus Send Failure: { \"message\" : \"${message.toLogString()}\"}")
-                                attemptToSend(message, messageTarget)
-                            }
-                            .onFailure { failure ->
-                                val logLine = "Fatal Proteus Send Failure: { \"message\" : \"${message.toLogString()}\"" +
-                                        " , " +
-                                        "\"errorInfo\" : \"${failure}\"}"
-                                logger.e(logLine)
-                            }
-                    }
-
-                    else -> {
-                        logger.e("Message Send Failure: { \"message\" : \"${message.toLogString()}\", \"errorInfo\" : \"${it}\" }")
-                        Either.Left(it)
-                    }
-                }
+                handleProteusError(it, "Send", message.toLogString()) { attemptToSend(message, messageTarget) }
             }, {
                 logger.i("Message Send Success: { \"message\" : \"${message.toLogString()}\" }")
                 Either.Right(it)
             })
+
+    /**
+     * Attempts to send a Proteus envelope without need to provide a specific conversationId
+     * Will handle the failure and retry in case of [ProteusSendMessageFailure].
+     */
+    private suspend fun tryBroadcastProteusEnvelope(
+        envelope: MessageEnvelope,
+        message: BroadcastMessage,
+        option: BroadcastMessageOption,
+        target: BroadcastMessageTarget
+    ): Either<CoreFailure, String> =
+        messageRepository
+            .broadcastEnvelope(envelope, option)
+            .fold({
+                handleProteusError(it, "Broadcast", message.toLogString()) { attemptToBroadcastWithProteus(message, target) }
+            }, {
+                logger.i("Message Broadcast Success: { \"message\" : \"${message.toLogString()}\" }")
+                Either.Right(it)
+            })
+
+    private suspend fun handleProteusError(
+        failure: CoreFailure,
+        action: String, // Send or Broadcast
+        messageLogString: String,
+        retry: suspend () -> Either<CoreFailure, String>
+    ) =
+        when (failure) {
+            is ProteusSendMessageFailure -> {
+                logger.w("Proteus $action Failure: { \"message\" : \"${messageLogString}\", \"errorInfo\" : \"${failure}\" }")
+                messageSendFailureHandler
+                    .handleClientsHaveChangedFailure(failure)
```

Comment:
What if the failure is not due to missing clients? do we want to execute this always on any kind of failure?